netket/hilbert/index/constrained.py

```diff
@@ -36,79 +53,166 @@ def compute_constrained_to_bare_conversion_table(
+        if self._n_max == 1:
+            return math.comb(self.size, self.n_particles)
+        else:
+            return self._lookup_table.n_states
 
     @property
-    def local_states(self):
-        return self._unconstrained_index._local_states
+    def _n_max(self):
+        return max(self.shape) - 1
 
     @property
-    def local_size(self) -> int:
-        return self._unconstrained_index.local_size
+    def size(self):
+        return len(self.shape)
+
+    def states_to_numbers(self, states: Array) -> Array:
+        return self._lookup_table.states_to_numbers(states)
+
+    def numbers_to_states(self, numbers: Array):
+        return self._lookup_table.numbers_to_states(numbers)
+
+    def all_states(self):
+        return self._lookup_table.all_states()
+
+    def _compute_all_states(self):
+        if self.n_particles == 0:
+            return jnp.zeros((1, self.size), dtype=jnp.int32)
+        c = jnp.repeat(
+            jnp.eye(self.size, dtype=jnp.int32),
+            np.array(self.shape) - 1,
+            axis=0,
+        )
+        combs = jnp.array(
+            list(itertools.combinations(np.arange(len(c)), self.n_particles))
+        )
+        all_states = c[combs].sum(axis=1, dtype=jnp.int32)
+        if (np.array(self.shape) > 1).any():
+            with jax.ensure_compile_time_eval():
+                all_states = jnp.unique(all_states, axis=0)
+        return jnp.asarray(all_states)
+
+    @struct.property_cached(pytree_node=True)
+    def _lookup_table(self) -> LookupTableHilbertIndex:
+        return LookupTableHilbertIndex(self._compute_all_states())
 
     @property
-    def _bare_numbers(self) -> np.ndarray:
-        """
-        Returns the conversion table between indices in the constrained space and
-        the corresponding unconstrained space.
-        """
-        if self.__bare_numbers is None:
-            self.__bare_numbers = compute_constrained_to_bare_conversion_table(
-                self._unconstrained_index, self._constraint_fn
-            )
-
-        return self.__bare_numbers
-
-    def states_to_numbers(self, states):
-        out = self._unconstrained_index.states_to_numbers(states)
+    def n_states_bound(self):
+        # upper bound on n_states, exact if n_max == 1
+        # number of combinations to check in _compute_all_states
+        return math.comb((np.array(self.shape) - 1).sum(), self.n_particles)
 
-        out = np.searchsorted(self._bare_numbers, out)
+    @property
+    def is_indexable(self):
+        # make sure we have less than than max_states to check in _compute_all_states
+        return _is_indexable(self.n_states_bound)
 
-        if np.max(out, initial=0) >= self.n_states:
-            raise RuntimeError(
-                "The required state does not satisfy " "the given constraints."
-            )
 
-        return out
+class SumConstrainedHilbertIndex(SumConstrainedHilbertIndexFock):
+    # shape is uniform, with same StaticRange on all sites
```

Comment:
add docstring saying why this exists. Maybe could go in the same file as `constraints.py`